src/utils/gcsApi.ts

```diff
@@ -0,0 +1,321 @@
+import * as fs from 'fs';
+import * as path from 'path';
+
+import {
+  Bucket,
+  Storage,
+  UploadOptions as GCSUploadOptions,
+} from '@google-cloud/storage';
+import * as ConsolaRaw from 'consola';
+import { isDryRun } from 'dryrun';
+
+import { logger as loggerRaw } from '../logger';
+import { reportError } from './errors';
+import { checkEnvForPrerequisite, RequiredConfigVar } from './env';
+
+let logger: typeof ConsolaRaw;
+
+const DEFAULT_MAX_RETRIES = 5;
+const DEFAULT_UPLOAD_METADATA = { cacheControl: `public, max-age=300` };
+
+const IS_DRY_RUN = isDryRun();
+
+/**
+ * Mapping between file extension regexps and the corresponding content type
+ * that will be set.
+ */
+const CONTENT_TYPES_EXT: Array<[RegExp, string]> = [
+  [/\.js$/, 'application/javascript; charset=utf-8'],
+  [/\.js\.map$/, 'application/json; charset=utf-8'],
+];
+
+/**
+ * Is this bucket being used as an artifact store or a target?
+ */
+export const enum BucketRole {
+  /** Artifact storage (used in both `prepare` and `publish`) */
+  STORE = 'artifact store',
+  /** Destination for `publish` */
+  TARGET = 'target',
+}
+
+/**
+ * Configuration options for the GCS bucket
+ */
+export interface GCSBucketConfig {
+  /** Bucket name */
+  bucketName: string;
+  /** ID of the project containing the bucket   */
+  projectId: string;
+  /** CGS credentials */
+  credentials: { client_email: string; private_key: string };
+  /** Role (is this being used as an artifact store or a target?) */
+  bucketRole: BucketRole;
+  /** Maximum number of retries after unsuccessful request */
+  maxRetries?: number;
+}
+
+/**
+ * Bucket path with associated parameters
+ */
+export interface DestinationPath {
+  /** Path inside the bucket to which files will be uploaded */
+  path: string;
+  /** Metadata to be associated with the files uploaded the path */
+  metadata: any;
+}
+
+/**
+ * Abstraction for a GCS bucket
+ */
+export class GCSBucket {
+  /** Bucket name */
+  public readonly bucketName: string;
+  /** CGS Client */
+  private readonly bucket: Bucket;
+
+  public constructor(config: GCSBucketConfig) {
+    const {
+      bucketName,
+      projectId,
+      credentials,
+      bucketRole,
+      maxRetries = DEFAULT_MAX_RETRIES,
+    } = config;
+
+    logger = loggerRaw.withScope(`[gcs ${bucketRole}]`);
+
+    this.bucketName = bucketName;
+    this.bucket = new Bucket(
+      new Storage({
+        credentials,
+        maxRetries,
+        projectId,
+      }),
+      bucketName
+    );
+  }
+
+  // TODO (kmclb) in readme deprecate CRAFT_GCS_CREDENTIALS_JSON/_PATH in favor of CRAFT_GCS_TARGET_CREDS_JSON/_PATH
+
+  /**
+   * Pulls GCS redentials out of the environment, where they can be stored either
+   * as a path to a JSON file or as a JSON string.
+   *
+   * @returns An object containing the credentials
+   */
+  public static getGCSCredsFromEnv = (
+    gcsRole: BucketRole
+  ): { [key: string]: string } => {
```

Comment:
Sure, I can do that.

Being new to typescript, I'm not sure where to strike the balance between the two extremes of no interfaces (easy! fewer lines of code the write or read! but also: why even use TS if that's what you're doing to do?) and making every one-off collection of properties in an object into an interface (ultimate type safety! every single object is documented! but also: a lot of work, and the strong likelihood of making your code so busy the reader can't tell what's important.).

Intuitively, I think I've been going with a heuristic something along the lines of:

- if it's used multiple times, maybe it should be an interface, so that you know all of those instances have the same structure
- if it's used as a method of passing data between modules - literally an _inter_face - (as is the case with `GCSBucketConfig`, for instance), almost certainly make it an interface, so you know the two modules are speaking the same language
- if it's only used once, internally, probably don't make it an interface, because you risk drowning out the interfaces from the first two cases (or at least making it difficult for the reader to tell whether it's _one of those_ or _one of these_)

but I'm sure there's more nuance to it. Any tips?

Reply:
What you wrote makes sense, but since you return an object and not a primitive type, I think it already makes sense to add a type for it: e.g., it helps you to catch things like misnamed or missing fields.
So I'd personally find the code easier to read if we had that interface.
You don't have to export it, of course.

Reply:
> but since you return an object and not a primitive type, I think it already makes sense to add a type for it: e.g., it helps you to catch things like misnamed or missing fields.

This seems to imply that any function/method which returns an object should have an associated interface. Is that generally the rule you go by?

In any case, added one here. Had to tweak the code a bit to make the compiler happy.

Reply:
I haven't written a lot of Typescript myself, but I feel like when we know the shape of the data we return (i.e. all the object attributes), why not having a type for it?

> because you risk drowning out the interfaces from the first two cases

I honestly have never experienced that. Do you have examples of this in the codebase?

Reply:
> > because you risk drowning out the interfaces from the first two cases
> 
> I honestly have never experienced that. Do you have examples of this in the codebase?

Not one in particular, no. More just an instinct to keep things tidy and following the idea of creating an interface for every one-off object that gets returned to its logical conclusion. Who knows - perhaps I'm worrying when I don't need to be, and in practice it's fine. ¯\_(ツ)_/¯